ruby sdk dev guide [draft] (#3560)
temporalio/documentation: docs/develop/ruby/core-application.mdx

Ruby workflows. This means there are several things workflows cannot do such as:

- Perform IO (network, disk, stdio, etc)
- Access/alter external mutable state
- Do any threading
- Do anything using the system clock (e.g. `Time.Now`)
- Make any random calls
- Make any not-guaranteed-deterministic calls

To prevent illegal workflow calls, a call tracer is put on the workflow thread that raises an exception if any illegal
calls are made.
Which calls are illegal is configurable in the worker options.

### Customize Workflow Type {#workflow-type}

Workflows have a Type that are referred to as the Workflow name.

The following examples demonstrate how to set a custom name for your Workflow Type.

You can customize the Workflow name with a custom name in a `workflow_name` class method call on the class.
The Workflow name defaults to the unqualified class name.

```ruby
class MyWorkflow < Temporalio::Workflow::Definition
  # Customize the name
  workflow_name :MyDifferentWorkflowName

  def execute(name)
    Temporalio::Workflow.execute_activity(
      MyActivity,
      { greeting: 'Hello', name: },
      start_to_close_timeout: 100
    )
  end
end
```

> ### Workflow Futures {#workflow-futures}

Reviewer: Not following. Futures have nothing to do with activities, they are a workflow utility for concurrent code running.

> ### Workflow Futures {#workflow-futures}

Reviewer: I see this as reference material, not top-to-bottom read material, and I expect hierarchical nav to represent that. Workflow utilities go with workflows, activity utilities go with activities, etc.

> ### Workflow Futures {#workflow-futures}

Reviewer 2: I think that's a good argument, but it's one we've been having here for years, and in practice, I think we wind up continually returning to the reality that people are approaching it 50-50 as reference material and top-to-bottom material. Notably, one place this PR *does* already deviate from the others in a way I didn't comment on today is by adding a Quickstart page to the top of the Ruby SDK section -- I thought this was a neat experiment that we'd build on -- in light of that finding. This is, imo, one of the only places where we're seriously violating an assumption that people might be top-to-bottom, hence why I'm drawing attention to this one.

`Temporalio::Workflow::Future` can be used for running things in the background or concurrently.
This is basically a safe wrapper around `Fiber.schedule` for starting and `Workflow.wait_condition` for waiting.

> Nothing uses futures by default, but they work with all workflow code/constructs.

Reviewer 2: ```suggestion
Futures are never used implicitly, but they work with all Workflow code and constructs.
```

end
fut3 = Temporalio::Workflow::Future.new do
  Temporalio::Workflow.execute_activity(MyActivity3, schedule_to_close_timeout: 300)
end

# Wait for them all to complete
Temporalio::Workflow::Future.all_of(fut1, fut2, fut3).wait

Temporalio::Workflow.logger.info("Got: #{fut1.result}, #{fut2.result}, #{fut3.result}")
```

Or, say, to wait on the first of 5 activities or a timeout to complete:

```ruby
# Start 5 activities
act_futs = 5.times.map do |i|
  Temporalio::Workflow::Future.new do
    Temporalio::Workflow.execute_activity(MyActivity, "my-arg-#{i}", schedule_to_close_timeout: 300)
  end
end
# Start a timer
sleep_fut = Temporalio::Workflow::Future.new { Temporalio::Workflow.sleep(30) }

# Wait for first act result or sleep fut
act_result = Temporalio::Workflow::Future.any_of(sleep_fut, *act_futs).wait
# Fail if timer done first
raise Temporalio::Error::ApplicationError, 'Timer expired' if sleep_fut.done?
# Print act result otherwise
Temporalio::Workflow.logger.info("Act result: #{act_result}")
```

There are several other details not covered here about futures, such as how exceptions are handled, how to use a setter

> proc instead of a block, etc. See the API documentation for details.

Reviewer 2: This should be a link, probably?

An Activity is a normal method execution that's intended to execute a single, well-defined action (either short or long-running), such as querying a database, calling a third-party API, or transcoding a media file.
An Activity can interact with world outside the Temporal Platform or use a Temporal Client to interact with a Temporal Service.
For the Workflow to be able to execute the Activity, we must define the [Activity Definition](/activity-definition).

You can develop an Activity Definition by creating a class that extends `Temporalio::Activity::Definition`.
To register a class as an Activity with a custom name, use the `activity_name` class method in the class definition.
Otherwise, the activity name is the unqualified class name.

```ruby
class MyActivity < Temporalio::Activity::Definition
  def execute(input)
    "#{input['greeting']}, #{input['name']}!"
  end
end
```

There is no explicit limit to the total number of parameters that an [Activity Definition](/activity-definition) may support.
However, there is a limit to the total size of the data that ends up encoded into a gRPC message Payload.

A single argument is limited to a maximum size of 2 MB.

> And the total size of a gRPC message, which includes all the arguments, is limited to a maximum of 4 MB.

Reviewer: I would like us to not recommend large payload codec, it is not a good pattern and has many problems. We admittedly lack a good blog post or docs on how best to handle this (can discuss internally on prioritizing that), but we should not encourage large payload codec.

> And the total size of a gRPC message, which includes all the arguments, is limited to a maximum of 4 MB.

Reviewer 2: Fine by me, in that case I revise my feedback to clarifying that you can usually just send strings or identifiers instead of needing to send large blobs as I/O

Also, keep in mind that all Payload data is recorded in the [Workflow Execution Event History](/workflow-execution/event#event-history) and large Event Histories can affect Worker performance.

> This is because the entire Event History could be transferred to a Worker Process with a [Workflow Task](/tasks#workflow-task).

Reviewer 2: Generally yes, and that's the model I've followed with https://github.com/temporalio/documentation/pull/3598, but you might be interested to know that I've gotten just as much feedback from people who don't like *that* model either, and there are a lot of clarity improvements we can make here that would not have gotten looked at if *not* for this porting opportunity. I am happy to add the other SDK page changes in this PR but I think people would then start talking about scope creep and blocking Ruby...

> This is because the entire Event History could be transferred to a Worker Process with a [Workflow Task](/tasks#workflow-task).

Reviewer: In general at least with code we don't like to mix refactoring and greenfield work because the refactoring gets lost in the details. If we have things we need to improve across SDK docs, it should be improved across SDK docs IMO separately from a single language migration.

Some SDKs require that you pass context objects, others do not.
When it comes to your application data—that is, data that is serialized and encoded into a Payload—we recommend that you use a single hash or object as an argument that wraps the application data passed to Activities.

> This is so that you can change what data is passed to the Activity without breaking a method signature.

Reviewer 2: Should have an example here using a Ruby object rather than just talking about it

Reviewer: We chose not to do this in other SDK docs. If we want to consistently provide this across SDKs, I think we should do so across SDKs and not make Ruby unique here.

Activity parameters are the method parameters of the `execute` method.
These can be any data type Temporal can convert.

> Technically this can be multiple parameters, but Temporal strongly encourages a single parameter containing all input fields.

Reviewer 2: This is redundant with the preceding paragraph

Reviewer: This is also the case in other SDK docs. Are you sure you only want to improve here and not there?

> ### Activity Concurrency and Executors {#activity-concurrency-and-executors}

Reviewer: This section is very much for implementation/definition of activities. The starting of activities is unrelated to this implementation aspect.

> ### Activity Concurrency and Executors {#activity-concurrency-and-executors}

Author: An admonition definitely works in this case.

The entire Execution history is transferred from the Temporal service to Workflow Workers when a Workflow state needs to recover.
A large Execution history can thus adversely impact the performance of your Workflow.

Therefore, be mindful of the amount of data you transfer through Activity invocation parameters or Return Values.
Otherwise, no additional limitations exist on Activity implementations.

To spawn an Activity Execution, use the `execute_activity` operation from within your Workflow Definition.

```ruby
class MyWorkflow < Temporalio::Workflow::Definition
  # Customize the name
  workflow_name :MyDifferentWorkflowName

  def execute(name)
    Temporalio::Workflow.execute_activity(
      MyActivity,
      { greeting: 'Hello', name: },
      start_to_close_timeout: 100
    )
  end
end
```

Activity Execution semantics rely on several parameters.
The only required value that needs to be set is either a [Schedule-To-Close Timeout](/encyclopedia/detecting-activity-failures#schedule-to-close-timeout) or a [Start-To-Close Timeout](/encyclopedia/detecting-activity-failures#start-to-close-timeout).
These values are set as keyword parameters.

> ### Get Activity Execution results {#get-activity-results}

Reviewer 2: I would remove this H4 and just keep the sentence beneath it.

Reviewer: Does this only apply to Ruby docs?

Reviewer 2: See https://github.com/temporalio/documentation/pull/3560#discussion_r2162118932, and tell me which method of improvement you would object to least :)

The Activity result is the returned from the `execute_activity` call.

## Run Worker Process {#run-worker-process}

> The [Worker Process](/workers#worker-process) is where Workflow Functions and Activity Functions are executed.

Reviewer 2: ```suggestion
The [Worker Process](/workers#worker-process) is where Workflow Functions and Activity Functions are actually executed.
In a Temporal application deployment, you ship and scale as many Workers as you need to handle the load of your Workflows and Activities.
```

Reviewer: Does this only apply to Ruby docs?

- Each [Worker Entity](/workers#worker-entity) in the Worker Process must register the exact Workflow Types and Activity Types it may execute.
- Each Worker Entity must also associate itself with exactly one [Task Queue](/task-queue).

> - Each Worker Entity polling the same Task Queue must be registered with the same Workflow Types and Activity Types.

Reviewer 2: These bullet points should be next to the little code sample down below and softened somewhat, otherwise it feels like we're giving people a lot of scary / arbitrary rules before they've seen how task queue assignment actually works.

Reviewer: Does this only apply to Ruby docs? Can we do a separate SDK docs review not specific to Ruby?

A [Worker Entity](/workers#worker-entity) is the component within a Worker Process that listens to a specific Task Queue.

> Although multiple Worker Entities can be in a single Worker Process, a single Worker Entity Worker Process may be perfectly sufficient.

Reviewer 2: I would cut this from this guide, doesn't make any sense to anyone who hasn't already gone deeper into scaling and deploys.

Author: Agreed

For more information, see the [Worker tuning guide](/develop/worker-performance).

A Worker Entity contains a Workflow Worker and/or an Activity Worker, which makes progress on Workflow Executions and Activity Executions, respectively.

> To develop a Worker, create a new `Temporalio::Worker` providing the Client and worker options which include Task Queue, Workflows, and Activities and more.

Reviewer 2: I think the word here shouldn't necessarily be "develop" because you don't need to do a lot of development to ship a Worker. Maybe something like:

```suggestion
Workers are implemented in each Temporal SDK, and can be deployed with just a bit of boilerplate.
To create a Worker, use `Temporalio::Worker.new()`, providing the Worker options which include Task Queue, Workflows, and Activities and more.
```
